components/script/dom/navigator.rs

```diff
@@ -320,9 +338,158 @@ impl NavigatorMethods<crate::DomTypeHolder> for Navigator {
+            return Err(Error::Type("Cannot parse URL".to_owned()));
+        };
+        if !matches!(url.scheme(), "http" | "https") {
+            return Err(Error::Type("URL is not http(s)".to_owned()));
+        }
+        let mut request_body = None;
+        // Step 4. Let headerList be an empty list.
+        let mut headers = HeaderMap::with_capacity(1);
+        // Step 5. Let corsMode be "no-cors".
+        let mut cors_mode = RequestMode::NoCors;
+        // Step 6. If data is not null:
+        if let Some(data) = data {
+            // Step 6.1. Set transmittedData and contentType to the result of extracting data's byte stream
+            // with the keepalive flag set.
+            let extracted_body = data.extract(&global, can_gc)?;
+            // Step 6.2. If the amount of data that can be queued to be sent by keepalive enabled requests
+            // is exceeded by the size of transmittedData (as defined in HTTP-network-or-cache fetch),
+            // set the return value to false and terminate these steps.
+            if let Some(total_bytes) = extracted_body.total_bytes {
+                if total_bytes > 64 * 1024 {
+                    return Ok(false);
+                }
+            }
+            // Step 6.3. If contentType is not null:
+            if let Some(content_type) = extracted_body.content_type.as_ref() {
+                // Set corsMode to "cors".
+                cors_mode = RequestMode::CorsMode;
+                // If contentType value is a CORS-safelisted request-header value for the Content-Type header,
+                // set corsMode to "no-cors".
+                if is_cors_safelisted_request_content_type(content_type.as_bytes()) {
+                    cors_mode = RequestMode::NoCors;
+                }
+                // Append a Content-Type header with value contentType to headerList.
+                //
+                // We cannot use typed header insertion with `mime::Mime` parsing here,
+                // since it lowercases `charset=UTF-8`: https://github.com/hyperium/mime/issues/116
+                headers.insert(
+                    header::CONTENT_TYPE,
+                    HeaderValue::from_str(content_type).unwrap(),
+                );
+            }
+            request_body = Some(extracted_body.into_net_request_body().0);
+        }
+        // Step 7.1. Let req be a new request, initialized as follows:
+        let request = RequestBuilder::new(None, url.clone(), global.get_referrer())
+            .mode(cors_mode)
+            .destination(Destination::None)
+            .policy_container(global.policy_container())
+            .insecure_requests_policy(global.insecure_requests_policy())
+            .has_trustworthy_ancestor_origin(global.has_trustworthy_ancestor_or_current_origin())
+            .method(http::Method::POST)
+            .body(request_body)
+            .origin(origin)
+            // TODO: Set keep-alive flag
```

Comment:
I've looked at `navigator.sendBeacon` before and if I recall correctly then the keep-alive is a fairly crucial part of the API 😅 . `sendBeacon` is usually called when the user navigates away from a page, to send analytics to the server for example and the keep-alive flag tells fetch to complete the request even if the originating browsing context no longer exists.

TLDR; I'm not sure how useful this API is without the flag ):

Reply:
The API will work for some cases, but indeed in those cases the request is not guaranteed to be sent given the lack of keep-alive. Still, I think it makes sense to land an initial implementation and implement keep-alive in a separate PR. That way, we keep things manageable? We did something similar with preload which is largely not working, but since it is quite a big feature we take smaller steps to get there.

Reply:
That seems reasonable but please put it behind a feature flag for now. Some websites do check if `navigator.sendBeacon` is available (for example MDN will log a warning about it), and merging this PR in it's current state will take away their ability to replace it with a blocking request.

Reply:
Done 👍